DESIGN.md

```diff
@@ -0,0 +1,370 @@
+
+`hero().fragments().heroDetails()`
+
+## Type conditions
+
+GraphQL allows you to specify type conditions on inline fragments to query type specific fields:
+
+```graphql
+query HeroDetails($episode: Episode) {
+  hero(episode: $episode) {
+    name
+    ... on Human {
+      height
+    }
+    ... on Droid {
+      primaryFunction
+    }
+  }
+}
+```
+
+One obvious way of mapping these would be to use subinterfaces for concrete object types:
+
+```java
+interface Hero {
+    String name();
+}
+
+interface Hero_Human extends Hero {
+    Float height();
+}
+
+interface Hero_Droid extends Hero {
+    String primaryFunction();
+}
+```
+
+This runs into the same complications with nested selections we've seen with fragments above however:
+
+```graphql
+query {
+  hero {
+    name
+    friends {
+      name
+    }
+    ... on Human {
+      height
+      friends {
+        appearsIn
+      }
+    }
+    ... on Droid {
+      primaryFunction
+      friends {
+        id
+      }
+    }
+  }
+}
+```
+
+```java
+interface Hero {
+  String name();
+  List<Friend> friends();
+  interface Friend {
+    String name();
+  }
+}
+
+interface Hero_Human extends Hero {
+  String name();
+  List<Friend> friends();
+  interface Friend extends Hero.Friend {
+    List<Episode> appearsIn();
+  }
+}
+
+interface Hero_Droid extends Hero {
+  String name();
+  List<Friend> friends();
+  interface Friend extends Hero.Friend {
+    String id();
+  }
+}
+```
+
+Again, we could use composition over inheritance here instead:
+
+```java
+interface Hero {
+  AsHuman asHuman();
+  interface AsHuman {
+    String name();
+    Float height();
+  }
+  AsHuman asDroid();
+  interface AsDroid {
+    String name();
+    String primaryFunction();
+  }
+}
+```
+
+These `as<>` properties are of course nullable or could return optionals.
+
+Named fragments can also specify type conditions:
+
+```graphql
+query TwoHeroes {
+  r2: hero {
+    ...HumanDetails
+    ...DroidDetails
+  }
+  luke: hero(episode: EMPIRE) {
+    ...HumanDetails
+    ...DroidDetails
+  }
+}
+
+fragment HumanDetails on Human {
+  name
+  height
+}
+
+fragment DroidDetails on Droid {
+  name
+  primaryFunction
+}
+```
+
+In a composition model, these map naturally onto nullable/optional properties:
+
+`hero().fragments().humanDetails()` (can be null)
+
+## Unique nested type names
+
+An unexpected complication in Java seems to be that nested types can never have the same simple name as an enclosing type. I'm not sure I understand the reasoning here (couldn't scoping rules or qualified names still differentiate between them?), but it is part of the language spec:
+
+> It is a compile-time error if a class has the same simple name as any of its enclosing classes or interfaces.
+http://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.1
+
+So unless we can find a way around this, we may have to give up on nesting and generate prefixed interface names ourselves.
```

Comment:
Wait @martijnwalraven, based on your statement `"Response names are guaranteed to be unique within a selection set"` I thought this wasn't a problem!?